### PR TITLE
[AutoDiff] Handle structs in direct adjoint value accumulation

### PR DESCRIPTION
**Pull request by jekbradbury**
This was a missing case in the adjoint emitter code.

### CODE REVIEW

**Comment by jekbradbury** [comment]
(resolves a compiler crasher in Transformer code that I haven't reduced)

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -4129,13 +4129,31 @@ AdjointEmitter::accumulateAdjointsDirect(AdjointValue &&lhs,
...
     case AdjointValueKind::Aggregate:
       SmallVector<AdjointValue, 8> newElements;
-      for (auto i : range(rhs.getNumAggregateElements())) {
-        auto lhsElt = builder.createTupleExtract(lhsVal.getLoc(), lhsVal, i);
-        auto rhsElt = rhs.takeAggregateElement(i);
-        newElements.push_back(accumulateAdjointsDirect(
-            makeConcreteAdjointValue(
-                ValueWithCleanup(lhsElt, lhsVal.getCleanup())),
-            std::move(rhsElt)));
+      auto lhsTy = lhsVal.getValue()->getType().getASTType();
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
      auto lhsTy = lhsVal.getSwiftType();
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -4129,13 +4129,31 @@ AdjointEmitter::accumulateAdjointsDirect(AdjointValue &&lhs,
...
+              makeConcreteAdjointValue(
+                  ValueWithCleanup(lhsElt, lhsVal.getCleanup())),
+              std::move(rhsElt)));
+        }
+      } else if (auto *structDecl = lhsTy->getStructOrBoundGenericStruct()) {
+        auto fieldIt = structDecl->getStoredProperties().begin();
+        for (unsigned i = 0; fieldIt != structDecl->getStoredProperties().end();
+             ++fieldIt, ++i) {
+          auto lhsElt = builder.createStructExtract(
+            lhsVal.getLoc(), lhsVal, *fieldIt);
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
              lhsVal.getLoc(), lhsVal, *fieldIt);
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 3]
```diff
@@ -4129,13 +4129,31 @@ AdjointEmitter::accumulateAdjointsDirect(AdjointValue &&lhs,
...
-        auto rhsElt = rhs.takeAggregateElement(i);
-        newElements.push_back(accumulateAdjointsDirect(
-            makeConcreteAdjointValue(
-                ValueWithCleanup(lhsElt, lhsVal.getCleanup())),
-            std::move(rhsElt)));
+      auto lhsTy = lhsVal.getValue()->getType().getASTType();
+      if (auto *tupTy = lhsTy->getAs<TupleType>()) {
+        for (auto idx : range(rhs.getNumAggregateElements())) {
+          auto lhsElt = builder.createTupleExtract(
+            lhsVal.getLoc(), lhsVal, idx);
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
              lhsVal.getLoc(), lhsVal, idx);
```

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (jekbradbury) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: That's what I tried at first, but `lhsVal` is a `ValueWithCleanup` and doesn't have a `getSwiftType` member.

Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: You can use `lhs.getSwiftType()`. But no big deal.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
Thanks!

**Pull request merged**